Archive/Draft/RFCxxxx-PS7-Bootstrapper.md

```diff
@@ -0,0 +1,256 @@
+
+#### Rust
+
+Rust has the benefit over C/C++ of having well defined contracts for resource management thus not having issues that persist in C/C++ code
+with regards to resource management which results in not only leaks but security issues.
+Resulting Rust binary code is very similar in size and performance as C/C++.
+Rust built code is supported in Windows.
+
+Rust uses the MS C runtime at execution, but we should statically link so that there aren't external dependencies that may be missing on the system.
+
+Need to evaluate what APIs are available on the smallest Windows SKU for downloading.
+
+This is the current preferred option.
+
+### Installer
+
+MSI has dependencies that are not available on Windows SKUs so would not be appropriate as the installer for the bootstrapper.
+Instead, we should look at using [Squirrel](https://github.com/Squirrel/Squirrel.Windows) which supports installing even if the product (existing version)
+is already in use.
+
+The bootstrapper would download the installer using Windows Update as the store to avoid issues with enterprises not allowing
+network access to other sites.
+The installer and binaries may need to be Windows signed instead of Microsoft signed.
+Packages of PS7 outside of this installer used by the boostrapper would continue to be Microsoft signed.
+
+### User Experience
+
+Since the initial bootstrapping will take time to download and install, the bootstrapper should have progress information and
+indicator during download and installation via stderr.
+Stderr is used instead of stdout so that user redirection of stdout only contains output from successful execution of their commandline.
+Use of stderr for informational messages is a convention used by native commands already.
+
+Should the install or download fail, the bootstrapper needs to output appropriate information and direct user to documentation on alternate
+install method and return a non-zero exit code.
+
+### Auto-updating
+
+Upon startup of the bootstrapper (following rules of [admin vs non-admin](#admin-vs-non-admins)) will detect if the installed
+version is installed by the bootstrapper and if so, if it is the latest stable version.
+
+If the version is older, then it will asynchronously download updated PS7 and continue with execution.
+A message will be emitted that on next startup PS7 will be updated.
+Upon next startup, if a newer version is downloaded, then the signature will be validated and installed providing
+progress information to the user.
+
+In general, auto-updates will be to a servicing release like 7.x.y to 7.x.y+1.
+However, when 7.x is out of support, then it should auto-update to 7.x+1.
+This is consistent with how our Microsoft Update support works to keep users patched and on a supported release.
+
+The auto-update applies regardless if PS7 is started interactively or via automation (explicitly using the `-noninteractive`, `-file`, or `-command` switch).
```

Comment:
Updating from 7.x to 7.x+1 can potentially break existing automation workload. Here are 2 real-word issue reported after the users moved to 7.2 from 7.1:
1. https://github.com/PowerShell/PowerShell/issues/16457 -- PowerCLI module was broken in 7.2 because of a breaking change (due to security requirement) in `Newtonsoft.Json`.
2. https://github.com/PowerShell/PowerShell/issues/16506 -- method resolution failed for a working script because new overloads were added to `System.Text.Json.JsonSerializer.Deserialize(...)`

Even if PowerShell itself can be backward regression free (best scenario case), the assemblies shipped with PowerShell can break user's existing workload, and there is nothing PowerShell can do to prevent that from happening. So, I think we should never automatically update from 7.x to 7.x+1. That should always be an explicit decision made by the user.

Reply:
It's a matter of opt-in or opt-out.  Secure by default means you have to opt-out of automatic updates.  I would expect folks who rely on automation to be experts vs folks who simply want to use a shell

Reply:
> I would expect folks who rely on automation to be experts vs folks who simply want to use a shell

I'm not necessarily arguing for or against automatic updates, but FWIW in my experience the opposite is true. At least on the Windows side, there's a much larger amount of less experienced folks who can cobble a script together than folks who are even a little comfortable in a shell.